modules/renter/uploadheap.go

```diff
@@ -235,14 +237,38 @@ func (r *Renter) buildUnfinishedChunks(f *file, hosts map[string]struct{}) []*un
 func (r *Renter) managedBuildChunkHeap(hosts map[string]struct{}) {
 	// Loop through the whole set of files and get a list of chunks to add to
 	// the heap.
-	id := r.mu.Lock()
+	id := r.mu.RLock()
+	goodForRenew := make(map[types.FileContractID]bool)
+	offline := make(map[types.FileContractID]bool)
 	for _, file := range r.files {
+		file.mu.RLock()
+		for cid := range file.contracts {
+			resolvedID := r.hostContractor.ResolveIDToPubKey(cid)
+			cu, ok := r.hostContractor.ContractUtility(resolvedID)
+			goodForRenew[cid] = ok && cu.GoodForRenew
+			offline[cid] = r.hostContractor.IsOffline(resolvedID)
+		}
+		file.mu.RUnlock()
+
 		unfinishedUploadChunks := r.buildUnfinishedChunks(file, hosts)
```

Comment:
This actually introduces a locking violation and a potential race condition. We can't call `build.UnfinishedChunks` without holding a write lock.